### PR TITLE
Rework the cancellation by using a custom CancelMonitor

### PR DESCRIPTION
**Pull request by henryju**
* able to propagate cancellation to downstream completable future
* able to watch shutdown of executor services

### CODE REVIEW

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/TextSearchIndex.java [hunk 1]
```diff
@@ -35,12 +35,12 @@
...
- * The result is sorted by score. The score of each term matches is the ratio of the term matches (1 for exact match), 
+ * The result is sorted by score. The score of each term matches is the ratio of the term matches (1 for exact match),
  * and the global score is the sum of the term's scores in the object divided by the total term frequency in the object.
- * 
+ * <p>
  * The generic type should properly implement equals and hashCode.
- * <b>An object cannot be indexed twice</b>. 
- * 
+ * <b>An object cannot be indexed twice</b>.
+ * <p>
```
Comment (damien-urruty-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/TextSearchIndex.java:43]: Should this be a closing tag?

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/branch/SonarProjectBranchTrackingService.java [hunk 2]
```diff
@@ -147,22 +147,22 @@ private String matchSonarProjectBranch(String configurationScopeId, SonarLintCan
...
     var mainBranchName = storedBranches.getMainBranchName();
-    var matchedSonarBranch = requestClientToMatchSonarProjectBranch(configurationScopeId, mainBranchName, storedBranches.getBranchNames(), cancelChecker);
+    var matchedSonarBranch = requestClientToMatchSonarProjectBranch(configurationScopeId, mainBranchName, storedBranches.getBranchNames(), cancelMonitor);
     if (matchedSonarBranch == null) {
       matchedSonarBranch = mainBranchName;
     }
-    if (cancelChecker.isCanceled()) {
+    if (cancelMonitor.isCanceled()) {
       throw new CancellationException();
     }
```
Comment (damien-urruty-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/branch/SonarProjectBranchTrackingService.java:156]: You could replace these 3 lines with `checkCanceled`

**Review by damien-urruty-sonarsource — APPROVED** [review]
LGTM

**Comment by damien-urruty-sonarsource** [comment]
There are also some SQ issues to fix

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/TextSearchIndex.java (hunk 1 above)
Comment (henryju) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/TextSearchIndex.java:43]: This is valid HTML 4 :) but fair enough, I will change it to HTML 5

**Pull request merged**